feat: replace ID with UID

sabhas opened the pull request:
BREAKING CHANGE: remove auto incremental ids from user, group and permissions and add a virtual uid property that returns string value of documents object id

## Issue

Closes #361.

While addressing #359, the auto-incrementing sequence counter used for
`User`, `Group`, and `Permission` IDs was found to be broken on Cosmos DB
(it set every ID to `1`). More generally, sequencing via a shared counter
document is not a good fit: it's a bottleneck on every
insert and a source of race conditions under concurrent writes.

## Intent

Replace auto-incremental numeric IDs with a string `uid` across `User`,
`Group`, and `Permission`, and make the API consistently expose `uid`
instead of `id` wherever these entities appear in a response.

## Implementation

- Added a virtual `uid` property to the `User`, `Group`, and `Permission`
  Mongoose models, returning `this._id.toString()`. No new field is stored
  and no data migration is required — every document already has `_id`,
  old and new alike.
- Removed the old sequencing mechanism entirely: the `Counter` model and
  `getSequenceNextValue` utility are gone, along with the bottleneck/race
  condition they caused.
- Updated all API responses that expose a user, group, or permission
  identifier — including login (`POST /SASLogon/login`), session
  (`GET /SASjsApi/session`), and the `user`/`group`/`permission` endpoints
  — to consistently return `uid`.
- Updated the web frontend (permission management, user profile, session
  restoration on page load) to read `uid` instead of `id`.

This is a **breaking change** for any existing client relying on numeric
IDs from this API — endpoints now return an opaque string identifier
instead.

## Checks

- [x] Code is formatted correctly (`npm run lint:fix`).
- [x] Any new functionality has been unit tested.
- [x] All unit tests are passing (`npm test`).
- [x] All CI checks are green.
- [x] Reviewer is assigned.

sabhas committed:
feat: replace ID with UID

BREAKING CHANGE: remove auto incremental ids from user, group and permissions and add a virtual uid property that returns string value of documents object id

sabhas committed:
chore: specs fixed

sabhas committed:
chore: fix specs

sabhas committed:
chore: fix specs

YuryShkoda commented on api/src/routes/api/spec/auth.spec.ts:
```diff
@@ -52,7 +52,7 @@ describe('auth', () => {
     const userInfo: InfoJWT = {
       clientId,
-      userId: user.id
+      userId: randomBytes(12).toString('hex')
```
`randomBytes(12).toString('hex')` should be a utility

YuryShkoda commented on api/src/routes/api/spec/web.spec.ts:
```diff
@@ -145,7 +145,7 @@ describe('web', () => {
       expect(res.body.loggedIn).toBeTruthy()
       expect(res.body.user).toEqual({
-        id: expect.any(Number),
+        id: expect.anything(),
```
I think it should be `any(String)`

YuryShkoda commented on web/src/context/appContext.tsx:
```diff
@@ -27,8 +27,8 @@ interface AppContextProps {
-  userId: number
-  setUserId: Dispatch<SetStateAction<number>> | null
+  userId: string
+  setUserId: Dispatch<SetStateAction<string>> | null
```
I think an `undefined` makes more sense than `null` here

YuryShkoda commented on web/src/context/appContext.tsx:
```diff
@@ -64,7 +64,7 @@ const AppContextProvider = (props: { children: ReactNode }) => {
   const [loggedIn, setLoggedIn] = useState(false)
   const [needsToUpdatePassword, setNeedsToUpdatePassword] = useState(false)
-  const [userId, setUserId] = useState(0)
+  const [userId, setUserId] = useState('')
```
I think an `undefined` should be a default value

YuryShkoda commented on web/src/utils/types.ts:
```diff
@@ -1,12 +1,12 @@
 export interface UserResponse {
-  id: number
+  uid: number
```
it should be a `string`

YuryShkoda commented on web/src/utils/types.ts:
```diff
@@ -1,12 +1,12 @@
 
 export interface GroupResponse {
-  groupId: number
+  uid: number
```
it should be a `string`

YuryShkoda commented on web/src/utils/types.ts:
```diff
@@ -17,7 +17,7 @@ export interface GroupDetailsResponse extends GroupResponse {
 
 export interface PermissionResponse {
-  permissionId: number
+  uid: number
```
it should be a `string`

YuryShkoda reviewed (changes requested):
see comments above

sabhas committed:
chore: addressed comments

sabhas committed:
chore: remove unused code

YuryShkoda committed:
Merge remote-tracking branch 'origin/main' into issue-361

YuryShkoda committed:
fix: update test fixtures for userId: string after ID->UID merge

Execution.spec.ts, processProgram.spec.ts and code.spec.ts didn't
exist when issue-361 (ID -> UID) branched, so they were written
against the old userId: number shape. Update them to match the
merged-in string-based uid now that main has been merged in.

YuryShkoda committed:
fix(api): return uid consistently from login and session endpoints

POST /SASLogon/login and GET /SASjsApi/session still returned the
old `id` field, while the rest of the ID->UID migration (#363)
standardized on `uid`. Not functionally broken - Mongoose provides
a built-in `id` virtual by default (_id.toHexString()) that happened
to resolve to the same value as the new `uid` virtual - but it's an
inconsistent public API surface, and relying on that coincidence
wasn't the intent of the migration.

Neither of these files was touched by any of issue-361's own
commits, so this predates the merge rather than being caused by it.

- web.ts: login response and session storage now source from
  user.uid explicitly
- session.ts: SessionResponse dropped its Omit<UserResponse, 'uid'>
  + id override in favor of just extending UserResponse
- verifyTokenInDB.ts: token-refresh path, same fix
- login.tsx / appContext.tsx: updated to read the corrected field

Verified with a real end-to-end request (genuine app boot, real
MongoDB, real CSRF handshake) - not just type-checking - to confirm
the actual HTTP response bodies carry uid, not id.
